### PR TITLE
deprecate unsafe_length for length

### PR DESCRIPTION
**Pull request by vtjnash**
This seems to be a fairly arbitrary case for throwing exceptions, when the user might often use this value in arithmetic afterwards (e.g. see `searchsorted`), which is not checked. It leads to awkward complexity in the API however, where it may be unclear which function to reach for, with no particular justification for why a particular usage is "safe". And it may inhibit optimization and performance due to the additional checks and error cases (and is not even entirely type-stable).

Performance seems to be basically identical:
```
julia> @btime length($(Ref(1:3:10))[])
  4.413 ns (0 allocations: 0 bytes)
julia> @btime length($(Ref(1:3:10))[])
  4.482 ns (0 allocations: 0 bytes)
```

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Very nice to eliminate these ugly functions. Triage is concerned about what checks we should have though. We discussed checking on construction whether a range will have a correct length, and moving the error there.

**Comment by vtjnash** [comment]
triage thought this was a reasonable change, but that we should keep the current version around as `checked_length` in the same module as the other `checked_math` functions

File: base/deprecated.jl [hunk 1]
```diff
@@ -240,4 +240,7 @@ end
 @deprecate cat_shape(dims, shape::Tuple{}, shapes::Tuple...) cat_shape(dims, shapes) false
 cat_shape(dims, shape::Tuple{}) = () # make sure `cat_shape(dims, ())` do not recursively calls itself
 
+@deprecate unsafe_indices(A) axes(A)
+@deprecate unsafe_length(r) length(r)
```
Comment (fredrikekre) [inline base/deprecated.jl:244]: These are currently not exported, so should be
```suggestion
@deprecate unsafe_indices(A) axes(A) false
@deprecate unsafe_length(r) length(r) false
```
I think.

**Comment by vtjnash** [comment]
@JeffBezanson this should be ready to go now. It removes checking for overflow in length for `Union{Int,UInt,Int64,UInt64,Int128,UInt128}`, and adds checking for all types in `Base.Checked.checked_length`. It is also a bit more consistent about the units of the computations, since `(b - a + s) / s` and `(b - a) / s + one` are mathematically equivalently, but could have different units, and we were using them interchangeably (this showed up on the added tests).

File: test/testhelpers/OffsetArrays.jl [hunk 2]
```diff
@@ -68,7 +68,7 @@ offset_coerce(::Type{I}, r::AbstractUnitRange) where I<:AbstractUnitRange{T} whe
 @inline Base.parent(r::IdOffsetRange) = r.parent
 @inline Base.axes(r::IdOffsetRange) = (Base.axes1(r),)
 @inline Base.axes1(r::IdOffsetRange) = IdOffsetRange(Base.axes1(r.parent), r.offset)
-@inline Base.unsafe_indices(r::IdOffsetRange) = (r,)
+#@inline Base.unsafe_indices(r::IdOffsetRange) = (r,) # why did this assume first(r.parent) == 1 previously??
```
Comment (JeffBezanson) [inline test/testhelpers/OffsetArrays.jl:71]: delete?

**Review by JeffBezanson — APPROVED** [review]

File: test/testhelpers/OffsetArrays.jl (hunk 2 above)
Comment (vtjnash) [reply test/testhelpers/OffsetArrays.jl:71]:
```suggestion
```

**Comment by giordano** [comment]
After this pull request
```julia
julia> using Unitful

julia> length(1u"m":2u"m":0u"m")
0 m
```

**Pull request merged**